File: android-pgs-plugin/src/AchievementsController.java
package org.godotengine.godot;

import android.app.Activity;
import android.content.Intent;
import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.games.Games;
import com.google.android.gms.tasks.OnSuccessListener;

public class AchievementsController {

    public static final int RC_ACHIEVEMENT_UI = 9003;

    private Activity activity;
    private ConnectionController connectionController;
    private GodotCallbacksUtils godotCallbacksUtils;

    public AchievementsController(Activity activity, ConnectionController connectionController, GodotCallbacksUtils godotCallbacksUtils) {
        this.activity = activity;
        this.connectionController = connectionController;
        this.godotCallbacksUtils = godotCallbacksUtils;
    }

    public void unlockAchievement(String achievementName) {
        GoogleSignInAccount googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity);
        if (connectionController.isConnected().first && googleSignInAccount != null) {
            Games.getAchievementsClient(activity, googleSignInAccount).unlock(achievementName);
            godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.ACHIEVEMENT_UNLOCKED, new Object[]{achievementName});
        } else {
            godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.ACHIEVEMENT_UNLOCKED_FAILED, new Object[]{achievementName});
        }
    }

    public void revealAchievement(String achievementName) {
        GoogleSignInAccount googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity);
        if (connectionController.isConnected().first && googleSignInAccount != null) {
            Games.getAchievementsClient(activity, googleSignInAccount).reveal(achievementName);
            godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.ACHIEVEMENT_REVEALED, new Object[]{achievementName});
        } else {
            godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.ACHIEVEMENT_REVEALED_FAILED, new Object[]{achievementName});
        }
    }

    public void incrementAchievement(String achievementName, int step) {
        GoogleSignInAccount googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity);
        if (connectionController.isConnected().first && googleSignInAccount != null) {
            Games.getAchievementsClient(activity, googleSignInAccount).increment(achievementName, step);
            godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.ACHIEVEMENT_INCREMENTED, new Object[]{achievementName});
        } else {
            godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.ACHIEVEMENT_INCREMENTED_FAILED, new Object[]{achievementName});
        }
    }

    public void showAchievements() {
        GoogleSignInAccount googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity);
        if (connectionController.isConnected().first && googleSignInAccount != null) {
            Games.getAchievementsClient(activity, googleSignInAccount)
                    .getAchievementsIntent()
                    .addOnSuccessListener(new OnSuccessListener<Intent>() {
                        @Override
                        public void onSuccess(Intent intent) {
                            activity.startActivityForResult(intent, RC_ACHIEVEMENT_UI);
                        }
                    });
        }
    }
}

File: app/src/main/java/io/cgisca/godot/gpgs/achievements/AchievementsController.kt
package io.cgisca.godot.gpgs.achievements

import android.app.Activity
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.games.Games
import com.google.android.gms.games.achievement.Achievement
import com.google.gson.Gson
import io.cgisca.godot.gpgs.ConnectionController
import io.cgisca.godot.gpgs.model.AchievementInfo

class AchievementsController(
    private val activity: Activity,
    private val achievementsListener: AchievementsListener,
    private val connectionController: ConnectionController
) {

    companion object {
        const val RC_ACHIEVEMENT_UI = 9003
    }

    fun unlockAchievement(achievementName: String) {
        val googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity)
        if (connectionController.isConnected().first && googleSignInAccount != null) {
            Games.getAchievementsClient(activity, googleSignInAccount).unlock(achievementName)
            achievementsListener.onAchievementUnlocked(achievementName)
        } else {
            achievementsListener.onAchievementUnlockingFailed(achievementName)
        }
    }

    fun revealAchievement(achievementName: String) {
        val googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity)
        if (connectionController.isConnected().first && googleSignInAccount != null) {
            Games.getAchievementsClient(activity, googleSignInAccount).reveal(achievementName)
            achievementsListener.onAchievementRevealed(achievementName)
        } else {
            achievementsListener.onAchievementRevealingFailed(achievementName)
        }
    }

    fun incrementAchievement(achievementName: String, step: Int) {
        val googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity)
        if (connectionController.isConnected().first && googleSignInAccount != null) {
            Games.getAchievementsClient(activity, googleSignInAccount).increment(achievementName, step)
            achievementsListener.onAchievementIncremented(achievementName)
        } else {
            achievementsListener.onAchievementIncrementingFailed(achievementName)
        }
    }

    fun setAchievementSteps(achievementName: String, steps: Int) {
        val googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity)
        if (connectionController.isConnected().first && googleSignInAccount != null) {
            Games.getAchievementsClient(activity, googleSignInAccount).setSteps(achievementName, steps)
            achievementsListener.onAchievementStepsSet(achievementName)
        } else {
            achievementsListener.onAchievementStepsSettingFailed(achievementName)
        }
    }

    fun showAchievements() {
        val googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity)
        if (connectionController.isConnected().first && googleSignInAccount != null) {
            Games.getAchievementsClient(activity, googleSignInAccount)
                .achievementsIntent
                .addOnSuccessListener { intent ->
                    activity.startActivityForResult(intent, RC_ACHIEVEMENT_UI)
                }
        }
    }

    fun loadAchievementInfo(forceReload: Boolean) {
        val googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity)
        if (connectionController.isConnected().first && googleSignInAccount != null) {
            Games.getAchievementsClient(activity, googleSignInAccount).load(forceReload)
                .addOnCompleteListener { task ->
                    val achievementData = task.result?.get()
                    if (task.isSuccessful && achievementData != null) {

                        val list = ArrayList<AchievementInfo>()
                        for(a in achievementData) {

                            val type = a.getType()
                            val current_steps = when(type == Achievement.TYPE_INCREMENTAL) {
                                true -> a.getCurrentSteps()
                                false -> null
                            }

                            val total_steps = when(type == Achievement.TYPE_INCREMENTAL) {
                                true -> a.getTotalSteps()
                                false -> null
                            }

                            list.add(AchievementInfo(
                                a.getAchievementId(),
                                a.getName(),
                                a.getDescription(),
                                a.getState(),
                                type,
                                current_steps,
                                total_steps,
                                a.getXpValue()
                            ))
                        }

                        achievementsListener.onAchievementInfoLoaded(Gson().toJson(list))
                    } else {
                        achievementsListener.onAchievementInfoLoadingFailed()
                    }
                }
        }
        else {
            achievementsListener.onAchievementInfoLoadingFailed()
        }
    }
}